pkg/policy/mapstate.go

```diff
@@ -195,29 +255,46 @@ func (keys MapState) addKeyWithChanges(key Key, entry MapStateEntry, adds, delet
 
 // deleteKeyWithChanges deletes a 'key' from 'keys' keeping track of incremental changes in 'adds' and 'deletes'.
 // The key is unconditionally deleted if 'cs' is nil, otherwise only the contribution of this 'cs' is removed.
-func (keys MapState) deleteKeyWithChanges(key Key, cs CachedSelector, adds, deletes MapState) {
+func (keys MapState) deleteKeyWithChanges(key Key, owner MapStateOwner, adds, deletes MapState) {
 	if entry, exists := keys[key]; exists {
-		if cs != nil {
+		if owner != nil {
 			// remove the contribution of the given selector only
-			if _, exists = entry.selectors[cs]; exists {
+			if _, exists = entry.owners[owner]; exists {
 				// Remove the contribution of this selector from the entry
-				delete(entry.selectors, cs)
-				// key is not deleted if other selectors still need it
-				if len(entry.selectors) > 0 {
+				delete(entry.owners, owner)
+				owner.RemoveDependent(keys, key)
```

Comment:
I'm curious why it makes sense to add the `RemoveDependent()` abstraction on the `MapStateOwner` for this implementation rather than against the `MapState`? I understand that the `owner` likely needs to be abstracted to be received into this function, but it's a bit odd that `MapStateOwner` interface only requires `RemoveDependent()` to be implemented (and not `AddDependent()`), and also that half of the implementers implement the function as a no-op (`CachedSelector`).

Reply:
The call sites will need to discriminate on the actual type of the `MapStateOwner` interface, but I agree, it will be clearer. Initially I thought that I could have `RemoveDependent()` only take a key, but now it also needs to take the map so it is better to make this a function on the map.